frame/examples/tasks-example/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{dispatch::DispatchResult, traits::Task};
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+use sp_runtime::DispatchError;
+
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+pub enum ExampleTask {
+	Increment,
+	Decrement,
+}
+
+pub trait PalletTask<T: Config>: Task {
+	type Config: frame_system::Config;
+	fn is_valid(&self, config: &Self::Config) -> bool;
+	fn run(&self, config: &Self::Config) -> Result<(), DispatchError>;
+}
+
+impl Task for ExampleTask {
+	type Enumeration = std::vec::IntoIter<ExampleTask>;
+
+	const TASK_INDEX: usize = 0;
+
+	fn enumerate() -> Self::Enumeration {
+		vec![ExampleTask::Increment, ExampleTask::Decrement].into_iter()
+	}
```

Comment:
Also separate confusion here, I am still a bit confused about whether `enumerate` is supposed to iterate over all possible _variants_ of a task, or something like all instances of the task that are currently queued. Is what I have here right?

Reply:
I believe it would be an iterator over all tasks, whereas `is_valid` would provide you with a boolean stating whether that particular variant is still valid.

Reply:
@sam0x17 Straight out of the issue:

> `tasks()` would ensure that the OCW/script could easily determine parameters which pass `condition()`.

So `tasks` enumerates all values which will make `condition` return `true`. This is needed for OCWs and other scripts which need to know what transactions to create and submit. How else could they craft the transactions with acceptable values of `task` otherwise?

Reply:
I thought I recalled answering this question before...

<img width="840" alt="image" src="https://github.com/paritytech/substrate/assets/138296/40f1d5ea-2942-440a-a176-66e429d435cb">